src/coord/rmr/rmr.c

```diff
@@ -638,22 +623,29 @@ MRIteratorCtx *MRIterator_GetCtx(MRIterator *it) {
 }
 
 MRReply *MRIterator_Next(MRIterator *it) {
-  void *p = MRChannel_Pop(it->ctx.chan);
-  // fprintf(stderr, "POP: %s\n", p == MRCHANNEL_CLOSED ? "CLOSED" : "ITER");
-  if (p == MRCHANNEL_CLOSED) {
-    return MRITERATOR_DONE;
+  return MRChannel_Pop(it->ctx.chan);
+}
+
+// Assumes no other thread is using the iterator, the channel, or any of the commands and contexts
+static void MRIterator_Free(MRIterator *it) {
+  for (size_t i = 0; i < it->len; i++) {
+    MRCommand_Free(&it->cbxs[i].cmd);
+  }
+  MRReply *reply;
+  while ((reply = MRChannel_UnsafeForcePop(it->ctx.chan))) {
+    MRReply_Free(reply);
   }
-  return p;
+  MRChannel_Free(it->ctx.chan);
+  rm_free(it->cbxs);
+  rm_free(it);
 }
 
-void MRIterator_WaitDone(MRIterator *it, bool mayBeIdle) {
-  if (mayBeIdle) {
-    // Wait until all the commands are at least idle (it->ctx.inProcess == 0)
-    while (MRIteratorCallback_GetNumInProcess(it)) {
-      usleep(1000);
-    }
-    // If we have no pending shards, we are done.
-    if (!it->ctx.pending) return;
+void MRIterator_Release(MRIterator *it) {
+  bool shouldFree = __atomic_test_and_set(&it->ctx.freeFlag, __ATOMIC_ACQUIRE);
+  if (!shouldFree) return;
+
+  // Both reader and writers are done with the iterator. No writer is in process.
+  if (it->ctx.pending) {
```

Comment:
Who is changing `pending`? Why shouldn't it be atomic?

Reply:
As the comment above says, at this point we are alone in the world so we don't need to do so